### PR TITLE
Support for network filesystems.

### LINKED ISSUE #25: VM created but it wont boot
Hi 
i am having a problem using this module correctly,  when i create a vm instance i am unable to actually get it to boot. I did check the xml and it seem to create storage in a completely different way which in might be ok but i noticed three things for me 

1. It wont boot permission error on storage create as pool instead of source file=. If manually fixed then i then hit the next problem 
2. No media is created by the module and it wont boot. If i manually fix xml then 
3. vnc part is set to /usr/bin/kvm and not to /usr/bin/kvm-spice. 

Anyway deleting the whole instance and keeping the same storage created by the module i am able to manually create instance  and boot with no problems. I am using just a basic example provided to create a vm in readme.

### LINKED ISSUE #35: Support for network disks
Hi,
I've been trying to massage in support for network filesystems - like ceph - as backing stock for instances.

https://github.com/goetzk/ansible-role-libvirt-vm/commit/69364760c7166757f2b87850d584049e0d2755f5

The problem I'm having with a minimal change addition is that the volume creation calls virt-volume.sh to do the volume creation - a step which isn't needed on a network disk.

Do you have a preferred way of circumventing this problem?
Karl.

### PR DESCRIPTION
**Pull request by goetzk**
In my case it would be ceph but hopefully there are enough switches included
that others could be used if desired.

I've tested this with RBD volumes.

Closes #35 .
Maybe Closes #25 .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -80,7 +80,7 @@ Role Variables
         - `name`: Name to associate with the volume being created; For `file` type volumes include extension if you would like volumes created with one.
         - `file_path`: Where the image of `file` type volumes should be placed; defaults to `libvirt_volume_default_images_path`
         - `device`: `disk` or `cdrom`
-        - `capacity`: volume capacity (can be suffixed with M,G,T or MB,GB,TB, etc) (required when type is `disk`)
+        - `capacity`: volume capacity, can be suffixed with k, M, G, T, P or E when type is `network` or MB,GB,TB, etc when type is `disk` (required when type is `disk`)
```
Comment (markgoddard) [inline README.md]: Looks like type is not documented. Could you add it?

Comment (markgoddard) [reply README.md]: Please also add auth and source.

File: tasks/volumes.yml [hunk 2]
```diff
@@ -27,11 +27,26 @@
     {% endif %}
     -a {{ ansible_check_mode }}
   with_items: "{{ volumes }}"
-  when: item.type | default(libvirt_volume_default_type) == 'volume'
+  when: "{{ item.device | default(libvirt_volume_default_device) == 'disk' }}"
```
Comment (markgoddard) [inline tasks/volumes.yml]: This needs to use type rather than device. We support type=volume with device=cdrom.

Comment (markgoddard) [reply tasks/volumes.yml]: type in ['network', 'volume'] ?

File: tasks/volumes.yml [hunk 3]
```diff
@@ -27,11 +27,26 @@
...
   environment: "{{ libvirt_vm_script_env }}"
   register: volume_result
   changed_when:
     - volume_result is success
     - (volume_result.stdout | from_json).changed | default(True)
   check_mode: False
   become: true
+
+# TODO: If the image already exists will this command reformat it while in use?
```
Comment (markgoddard) [inline tasks/volumes.yml]: Does this need to be done before the libvirt volume is created?

File: templates/vm.xml.j2 [hunk 4]
```diff
@@ -35,7 +35,18 @@
       <driver name='qemu' type='{{ volume.format | default(libvirt_volume_default_format) }}'/>
       {% if volume.type | default(libvirt_volume_default_type) == 'file' %}
       <source file='{{ volume.file_path |default(libvirt_volume_default_images_path) }}/{{ volume.name}}'/>
-      {% else %}
+      {% elif volume.type is defined and volume.type == 'network' %}
```
Comment (markgoddard) [inline templates/vm.xml.j2]: `volume.type | default(libvirt_volume_default_type) == 'network'`

File: templates/vm.xml.j2 [hunk 5]
```diff
@@ -35,7 +35,18 @@
...
+        {% if volume.auth.name %}
```
Comment (markgoddard) [inline templates/vm.xml.j2]: is defined?

File: templates/vm.xml.j2 [hunk 6]
```diff
@@ -35,7 +35,18 @@
...
+        <auth username='{{ volume.auth.name }}'>
+          <secret type='{{ volume.auth.type }}' uuid='{{ volume.auth.token }}'/>
+        </auth>
+        {% endif %} {# End volume.auth.name check #}
+        {% if volume.source.name %}
```
Comment (markgoddard) [inline templates/vm.xml.j2]: is defined?

File: tasks/volumes.yml (hunk 2 above)
Comment (markgoddard) [reply tasks/volumes.yml]: (see https://github.com/stackhpc/ansible-role-libvirt-vm/issues/45 for context)

File: README.md (hunk 1 above)
Comment (goetzk) [reply README.md]: I'll add some words for auth and source including references to their sub values.

File: tasks/volumes.yml (hunk 2 above)
Comment (goetzk) [reply tasks/volumes.yml]: I've changed this line to use the `in` idiom, testing to follow.

File: tasks/volumes.yml (hunk 3 above)
Comment (goetzk) [reply tasks/volumes.yml]: I believe this behaves but still need to explicitly double check.

File: templates/vm.xml.j2 (hunk 4 above)
Comment (goetzk) [reply templates/vm.xml.j2]: I've updated this one

File: templates/vm.xml.j2 (hunk 6 above)
Comment (goetzk) [reply templates/vm.xml.j2]: added.

File: README.md (hunk 1 above)
Comment (goetzk) [reply README.md]: This is done in the latest commit/s

File: tasks/volumes.yml (hunk 3 above)
Comment (goetzk) [reply tasks/volumes.yml]: To answer your question: the volume can be created any time before the instance is.

To answer the TODO note: volumes are not reformatted on subsequent runs.

File: tasks/volumes.yml [hunk 7]
```diff
@@ -27,11 +27,25 @@
...
+- name: Ensure the VM network volumes exist
+  command: qemu-img create -f {{ item.source.protocol }} {{ item.source.protocol }}:{{ item.source.name }} {{ item.capacity }}
+  with_items: "{{ volumes }}"
+  when: item.type | default(libvirt_volume_default_type) == 'network'
+  register: volume_result_network
+  # 0 is OK, 1 is an existing image
+  failed_when: volume_result_network.rc >= 2
+  changed_when:
+    - volume_result_network is success
+    - (volume_result_network.stdout | from_json).changed | default(True)
```
Comment (markgoddard) [inline tasks/volumes.yml]: Does this command output JSON? Don't we just want this:

`volume_result_network.rc == 1`

File: templates/vm.xml.j2 (hunk 5 above)
Comment (markgoddard) [reply templates/vm.xml.j2]: If the variable is undefined this will fail.

`{% if volume.auth.name is defined %}`

File: README.md [hunk 8]
```diff
@@ -75,12 +75,15 @@ Role Variables
...
       defined with the following dict:
+        - `type`: Type of the volume. Options include `volume`, `file` and `network`.
         - `pool`: Name or UUID of the storage pool from which the volume should be
-          allocated.
+          allocated. Required when `type` is `volume`.
         - `name`: Name to associate with the volume being created; For `file` type volumes include extension if you would like volumes created with one.
         - `file_path`: Where the image of `file` type volumes should be placed; defaults to `libvirt_volume_default_images_path`
         - `device`: `disk` or `cdrom`
-        - `capacity`: volume capacity (can be suffixed with M,G,T or MB,GB,TB, etc) (required when type is `disk`)
+        - `capacity`: volume capacity, can be suffixed with k, M, G, T, P or E when type is `network` or MB,GB,TB, etc when type is `file` (required when type is `file`)
```
Comment (markgoddard) [inline README.md]: Looks like this is also required when type is network.

File: templates/vm.xml.j2 [hunk 9]
```diff
@@ -35,7 +35,18 @@
...
       {% if volume.type | default(libvirt_volume_default_type) == 'file' %}
       <source file='{{ volume.file_path |default(libvirt_volume_default_images_path) }}/{{ volume.name}}'/>
-      {% else %}
+      {% elif volume.type | default(libvirt_volume_default_type) == 'network' %}
+        {% if volume.auth.name %}
+        <auth username='{{ volume.auth.name }}'>
+          <secret type='{{ volume.auth.type }}' uuid='{{ volume.auth.token }}'/>
+        </auth>
+        {% endif %} {# End volume.auth.name check #}
+        {% if volume.source.name is defined %}
```
Comment (markgoddard) [inline templates/vm.xml.j2]: Is the source mandatory? We use it earlier unconditionally with qemu-img.

Comment (goetzk) [reply templates/vm.xml.j2]: If the type is network source is required so i'll adjust that check.

File: README.md (hunk 8 above)
Comment (goetzk) [reply README.md]: Updated in latest commit

File: templates/vm.xml.j2 (hunk 5 above)
Comment (goetzk) [reply templates/vm.xml.j2]: Updated in latest changes

File: tasks/volumes.yml (hunk 7 above)
Comment (goetzk) [reply tasks/volumes.yml]: I need to investigate this, other feedback has been incorporated.

Comment (goetzk) [reply tasks/volumes.yml]: I've checked and qemu-img is not outputting json; I've applied your suggested change.

**Comment by goetzk** [comment]
The latest commit sits better on this stack than master so I've added it here.

File: tasks/volumes.yml [hunk 10]
```diff
@@ -27,11 +27,26 @@
     {% endif %}
     -a {{ ansible_check_mode }}
   with_items: "{{ volumes }}"
-  when: item.type | default(libvirt_volume_default_type) == 'volume'
+  # File added since volumes can specify a path to a file rather than using volumes
+  when: item.type | default(libvirt_volume_default_device) in ['volume', 'network', 'file']
```
Comment (markgoddard) [inline tasks/volumes.yml]: Are you sure you want to do this for file type volumes?

**Comment by goetzk** [comment]
Thanks mark,
I'll rebase my local branch and see whats left over but along with the changes in #58 this is looking good.

**Pull request merged**